File: src/main/java/miniDB/index/bptree/MainDataConfiguration.java
package miniDB.index.bptree;

import com.lss233.minidb.engine.storage.StorageType;
import com.lss233.minidb.exception.MiniDBException;

import java.util.ArrayList;

public class MainDataConfiguration extends Configuration {
    public int nValidPointerInFreePage;
    public MainDataConfiguration(ArrayList<StorageType> types, ArrayList<Integer> sizes, ArrayList<Integer> colIDs) throws MiniDBException {
        super(0, types, sizes, colIDs);
        // pad keysize to multiples of 16
        int BASE = 16;
        int tmpSize = keySize;
        tmpSize += 8; // space for `RowID`
        if(tmpSize % BASE != 0)
        {
            tmpSize += BASE - (tmpSize % BASE);
        }
        pageSize = Math.max(32, tmpSize);
        nValidPointerInFreePage = pageSize / 8 - 1;
    }
}

File: src/main/kotlin/com/lss233/minidb/engine/index/bptree/MainDataConfiguration.kt
package com.lss233.minidb.engine.index.bptree

import java.lang.reflect.Type

class MainDataConfiguration(types: ArrayList<Type>, sizes: ArrayList<Int>, colIDs: ArrayList<Int>):
    Configuration(0, types, sizes, colIDs) {

    var nValidPointerInFreePage: Int

    init {
        // pad keysize to multiples of 16
        val BASE = 16
        var tmpSize = keySize
        tmpSize += 8 // space for `RowID`
        if (tmpSize % BASE != 0) {
            tmpSize += BASE - tmpSize % BASE
        }
        pageSize = 32.coerceAtLeast(tmpSize)
        nValidPointerInFreePage = pageSize / 8 - 1
    }
}
